### ISSUE TITLE
Consider using github api to get the default branch 

### ISSUE THREAD

**Issue by weshaggard** [Azure/azure-sdk-tools#1570]
Currently we do some hacky parsing of the git remote show command https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/pipelines/templates/steps/set-default-branch.yml#L6

Given most if not all of our repo's are on github we could consider using the github api https://docs.github.com/en/rest/reference/repos#get-a-repository which has default_branch as property on the repository information.

**Comment by mitchdenny** [comment]
We may have an issue here with our ```-pr``` repositories. If we have any parts of the pipeline that run under the public team project we won't have access to a token we can use to access the private repo. To mitigate this we could make the assumption that if you are executing in a repo with is suffixed with ```-pr``` that you can make a request to the non-pr repo has the same default branch.

Thoughts?

**Comment by weshaggard** [comment]
For the private repo's I'd like to see if we can use the ACCESS_TOKEN that is already need to do the cloning before we go and special case "-pr" repos.

**Comment by mitchdenny** [comment]
Ended up finding out a way of doing this (mentioned it in the associated PR). Basically we set ```persistCredentials: true``` on the checkout operation, and then we can read the GitHub token out of the Git configuration (using ```git config --list``` and some filtration. The downside is that this leaves the Git token around on disk during the build.

On one hand that isn't good, but on the other anyone who can submit a PR that auto triggers can put this in there even if we decide not to do it (so no real security loss - just feels uncomfortable). Just another thing to check if someone outside the team submits a PR.
